src/hub/action_request.ts

```diff
@@ -118,6 +133,92 @@ export class ActionRequest {
   webhookId?: string
   lookerVersion: string | null = null
 
+  /** `stream` creates and manages a stream of the request data
+   *
+   * ```ts
+   * let prom = await request.stream(async (readable) => {
+   *    return myService.uploadStreaming(readable).promise()
+   * })
+   * ```
+   *
+   * Streaming generally occurs only if Looker sends the data in a streaming fashion via a push url,
+   * however it will also wrap non-streaming attachment data so that actions only need a single implementation.
+   *
+   * @returns The return value of the `callback` function. This is useful for returning
+   * a promise from the `callback` function.
+   * @param callback A function will be caled with a Node.js `Readable` object.
+   * The readable object represents the streaming data.
+   */
+  stream<T>(callback: (readable: Readable) => T): T {
+    const url = this.scheduledPlan && this.scheduledPlan.downloadUrl
+    const stream = new PassThrough()
+    const returnVal = callback(stream)
+    if (url) {
+      httpRequest.get(url).pipe(stream)
+    } else {
+      if (this.attachment && this.attachment.dataBuffer) {
+        stream.end(this.attachment.dataBuffer)
+      } else {
+        throw new Error(
+          "startStream was called on an ActionRequest that does not have" +
+          "a streaming download url or an attachment. Ensure usesStreaming is set properly on the action.")
+      }
+    }
+    return returnVal
+  }
+
+  /**
+   * A streaming helper for the "json" data format. It handles automatically parsing
+   * the JSON in a streaming fashion. You just need to implement a function that will
+   * be called for each row.
+   *
+   * @returns A promise that will be resolved when streaming is complete.
+   * @param onRow A function that will be called for each streamed row, with the row as the first argument.
+   */
+  async streamJson(onRow: (row: { [fieldName: string]: any }) => void) {
+    return new Promise<void>((resolve) => {
+      this.stream((readable) => {
+        oboe(readable)
+          .node("![*]", (row) => {
```

Comment:
is it worth calling out what the format structure means?

Reply:
i think its complex enough that you have to look at the oboe docs anyway, which should be relatively obvious